contracts/token/ERC20/extensions/ERC4626.sol

```diff
@@ -0,0 +1,187 @@
+
+        return shares;
+    }
+
+    /** @dev See {IERC4262-mint} */
+    function mint(uint256 shares, address receiver) public virtual override returns (uint256) {
+        require(shares <= maxMint(receiver), "ERC4626: mint more then max");
+
+        address caller = _msgSender();
+        uint256 assets = previewMint(shares);
+
+        // if _asset is ERC777, transferFrom can call reenter BEFORE the transfer happens through
+        // the tokensToSend hook, so we need to transfer before we mint to keep the invariants.
+        SafeERC20.safeTransferFrom(_asset, caller, address(this), assets);
+        _mint(receiver, shares);
+
+        emit Deposit(caller, receiver, assets, shares);
+
+        return assets;
+    }
+
+    /** @dev See {IERC4262-withdraw} */
+    function withdraw(
+        uint256 assets,
+        address receiver,
+        address owner
+    ) public virtual override returns (uint256) {
+        require(assets <= maxWithdraw(owner), "ERC4626: withdraw more then max");
+
+        address caller = _msgSender();
+        uint256 shares = previewWithdraw(assets);
+
+        if (caller != owner) {
+            _spendAllowance(owner, caller, shares);
+        }
+
+        // if _asset is ERC777, transfer can call reenter AFTER the transfer happens through
+        // the tokensReceived hook, so we need to transfer after we burn to keep the invariants.
+        _burn(owner, shares);
+        SafeERC20.safeTransfer(_asset, receiver, assets);
+
+        emit Withdraw(caller, receiver, owner, assets, shares);
+
+        return shares;
+    }
+
+    /** @dev See {IERC4262-redeem} */
+    function redeem(
+        uint256 shares,
+        address receiver,
+        address owner
+    ) public virtual override returns (uint256) {
+        require(shares <= maxRedeem(owner), "ERC4626: redeem more then max");
+
+        address caller = _msgSender();
+        uint256 assets = previewRedeem(shares);
+
+        if (caller != owner) {
+            _spendAllowance(owner, caller, shares);
+        }
+
+        // if _asset is ERC777, transfer can call reenter AFTER the transfer happens through
+        // the tokensReceived hook, so we need to transfer after we burn to keep the invariants.
+        _burn(owner, shares);
+        SafeERC20.safeTransfer(_asset, receiver, assets);
+
+        emit Withdraw(caller, receiver, owner, assets, shares);
+
+        return assets;
+    }
+
+    /**
+     * @dev Internal convertion function (from assets to shares) with support for rounding direction
+     *
+     * Will revert if assets > 0, totalSupply > 0 and totalAssets = 0. That corresponds to a case where any asset
+     * would represent an infinite amout of shares.
+     */
+    function _convertToShares(uint256 assets, Math.Rounding direction) internal view virtual returns (uint256 shares) {
+        uint256 supply = totalSupply();
+        return
+            (assets == 0 || supply == 0)
```

Comment:
This is here for the case where `totalAssets() == 0`.

In that case, the exchange rate from assets to shares is not properly defined:
- if the vault is empty (supply == 0) then we fall back to the decimal rate → ok
- if the vault is not empty (supply > 0), every asset is work an infinite amount of shares
  - in that case, I do `0 * infinity = 0` (using the fallback to decimals as an escape) to avoid the division by 0.
  
  The whole thing would be rewritten
  
  ```
        uint256 supply = totalSupply();
        return
            (supply == 0)
                ? assets.mulDiv(10**decimals(), 10**_asset.decimals(), direction)
                : (assets == 0)
                    ? 0
                    : assets.mulDiv(supply, totalAssets(), direction);
  ```